### DISCUSSION TITLE
How to avoid Tasmota loses stored data

### DISCUSSION THREAD

**Discussion by ChrisDupu** [arendst/Tasmota#17825] [Q&A]
Bonjour,

I'm using Tasmota 12.1.1.6 with a socket-plug to count what power a small solar system generated. The possibilities to "analyse" the genereted power ("Energy today/yesterday/total") are really rudimentary, but OK for me. Unfortunately they are not reliable:

- when I restart my socket via the web interface, button "Restart" everything is fine and it comes back with all the values like they have been before the restart
- when I cycle the power, means unplug the socket completely, the energy values get lost and it comes back with some random/old values!

So this is not helpful as the main power can fail spontaneously from time to time.

My question: can this be automated somehow? Means is there a possibility to auto-save the current values e.g. every 6 hours?

Merci!

**Comment by barbudor** [comment]
Bonjour

In order to avoid wearing flash, tasmota do not write continuously in flash but only
- when a setting is changed
- on requested restart (red button on GUI or restart command in the console, mqtt)
- at midnight

If you want more frequent save, you can use a rule and use command 'savedata'

This question is asked about once a month and there should already be quite a few examples of rules.
Exemple
```
Rule1 on time#minute|360 do savedata endon
```

**Comment by ChrisDupu** [comment]
Thanks for feedback. I checked this, but the rule "at midnight" does not work. After doing a power-cycle right now, the shown statistics data are reset to values which are not the ones from yesterday but something more old. Or some random values, can't say that exactly.

**Reply by arendst** [reply]
Upgrade to at least v12.3.1 to have a better experience.

**Reply by ChrisDupu** [reply]
Merci, I'll try the version!

Is there by any chance the possibility to have some more detailed statistics in the new version? E.g. the sum of power of each of the twelve months? This would be a really nice feature for solar power evaluation.

**Reply by barbudor** [reply]
That's something you should do in a back-end
Tasmota runs on small processors, especially ESP8266 on socket plugs, that do not have much resources, especially flash.
If you don't want to run a back-end, move to a device with an ESP32 and you can add such feature using the Berry language to save data in the flash file system

On ESP8266 you could use Scripting Language but the 1MB flash on a socket plug do not leave enough room to save such information

**Reply by ChrisDupu** [reply]
So after upgrading the software to version 12.3.1.5 the "Flash write count" has not changed during the night. This means, midnight auto-save did not work?

About the lack of memory: the information page says "Free Program Space 327 KB" and "Free Memory 25,9 KB". Isn't this more than enough space for 12 simple values storing?

**Reply by ChrisDupu** [reply]
Any news on this?

**Reply by barbudor** [reply]
If tasmota doesn't save data may be it is because you've set 'savedata 0'
Check the return of 'savedata' without arguments

**Reply by ChrisDupu** [reply]
This "Savedata" is "off". But: the manual from https://tasmota.github.io/docs/Commands/#management says, savedata can be used to set a period of time to save changed parameters. What I want to get, is the parameters to be saved on midnight as @arendst explained above

**Reply by pkkrusty** [reply]
Seems like you could hack something together using vars and mems...I'll have to think on it a bit.

**Reply by arendst** [reply]
See your own issue. You probably misinterpreted the manual. What it tries to say is that, when savedata is 1, it checks any changes in the settings and if there is one it saves it to flash. It doesn't say it writes to flash every second.

At midnight the RTC values are transferred to the settings area which in turn starts the write to flash if savedata is at least 1.

**Reply by pkkrusty** [reply]
Here's a ruleset that more or less does what you want (I think). May have to play with the Energy syntax.

```
rule1
on time#minute do counter1 %utctime% endon
on time#minute do counter1 -1672531200 endon
on tele-counter#c1 do var13 %value% endon
on var13#state do var15=var13/86400 endon
on tele-energy#today do var14 %value% endon
on time#minute=1439 do if (var15>=0 and var15 <31) event t1=1; mem2 0 endif endon
on time#minute=1439 do if (var15>=31 and var15 <59) event t1=2; mem3 0 endif endon
on time#minute=1439 do if (var15>=59 and var15 <90) event t1=3; mem4 0 endif endon
on time#minute=1439 do if (var15>=90 and var15 <120) event t1=4; mem5 0 endif endon
on time#minute=1439 do if (var15>=120 and var15 <151) event t1=5; mem6 0 endif endon
on time#minute=1439 do if (var15>=151 and var15 <181) event t1=6; mem7 0 endif endon
on time#minute=1439 do if (var15>=181 and var15 <212) event t1=7; mem8 0 endif endon
on time#minute=1439 do if (var15>=212 and var15 <243) event t1=8; mem9 0 endif endon
on time#minute=1439 do if (var15>=243 and var15 <273) event t1=9; mem10 0 endif endon
on time#minute=1439 do if (var15>=273 and var15 <304) event t1=10; mem11 0 endif endon
on time#minute=1439 do if (var15>=304 and var15 <334) event t1=11; mem12 0 endif endon
on time#minute=1439 do if (var15>=334 and var15 <365) event t1=12; mem1 0 endif endon

rule2
on event#t1 do mem%value%=mem%value%+var14 endon
```
Need to compile with `#USE_EXPRESSION` and `#USE_IF_STATEMENT` and also configure two Counters so you can use their 31-bit space.
Each Jan-1, you would need to adjust the epoch time that is subtracted, and in 2024 would need to adjust the days due to the leap-year.
So, kind of hacky, but gives you persistent accumulated Energy totals by month. Probably someone could improve on it. Also, with a power cut, you will lose whatever accumulated day total you had.

As others have said, scripting/berry would be simpler.

**Reply by ChrisDupu** [reply]
@arendst The manual from https://tasmota.github.io/docs/Commands/#management says:
"SaveData |...|1 = save parameter changes every second (default)"

This is definitely NOT what I am looking for! Saving the parameters **every** second is way too often and for sure would kill the flash quite fast. Or should this mean "1 = save parameter one second after a change"? This is not what the manual says. And this does not necessarily mean it performs a save at midnight.

@barbudor I added a rule "Rule1 on time#minute|1000 do savedata endon" but this does not save the data properly. With this rule the flash write count is increaded regularly, but the energy values are NOT saved. After a power failure "Today energy" and "Yester day energy" are set to 0 and "Total energy" falls back to the value I had on last manual reboot.

@pkkrusty thank you for the rules - I'll try them :-)

**Reply by arendst** [reply]
Ask yourselves what the difference is between "saving every second" and "saving changes every second". Also notice that there are no changes to the settings if energy values change. Only at midnight or when executing command ``restart 1`` the energy values are stored in the settings area which in turn will save the changes.

If I were you I would be very advanterous and forget the manual but execute command ``SaveData 1`` and take a break. Wait a few days and look at your favorite flash write count. You'll be surprised. After that re-read the manual and see the eureka moment.

No need for rules.
